### PR TITLE
[MC-45] Create Logic Component Input

### LINKED ISSUE #45: [Feature] Create Logic Component Input
**Description**
I want a class to be created that represents the text Input that will be used as the default for the entire site, being possible to instantiate the object for each component.

**Additional context**
const imp1 = new Ipunt('btn query 1');
const imp2 = new Ipunt('btn query 2');

**Acceptance Criteria**
- The component must have its class
- The component needs to keep its state (disabled/enabled)
- The component must have a get/set value
- The component must have validation (Email)
- The component must have required (true|false)
- The component is expected to have tests

### PR DESCRIPTION
**Pull request by RodrigoSoaresNascimento**
# Changes Summary

- Add get/set Text Input;
- Add get/set Email Input;
- Add Text and Email input validations (email validation tests have been added in google format fulano@gmail.com similar and for texts anything other than empty is a valid text);
- Add get/set valid, enabled and disabled to text and email;
- Add verify into valid, enabled and disabled to text and email;
- Add tests to all this new methods.
- Add a new documents text-input.components.js and  text-input.components.tests.js

# How Has This Been Tested?

To test all methods, i created new obects using them, checked in the localhost log to see these returns.

### CODE REVIEW

File: src/assets/scss/checkbox.scss [hunk 1]
```diff
@@ -1,20 +1,26 @@
-fieldset {
-    border: none;
-    display: flex;
-    flex-direction: column;
-    margin: 20px;
-    align-items: flex-start;
-    min-width: 400px;
+.container-checkbox-component{
```
Comment (EvertonDuarteGuimaraes) [inline src/assets/scss/checkbox.scss]: That's a really big class name to deal with simple styles, maybe you should do something like:
class="checkbox-container checkbox-group checkbox-filled"...
Like this you can optimize your SCSS to organize it, check this out:

.checkbox- {
....&container {
........// styles for container
....}

....&group {
........// styles for group
....}

....&filled {
........// styles for filled
....}
}

File: src/assets/scss/checkbox.scss [hunk 2]
```diff
@@ -1,20 +1,26 @@
...
+        display: flex;
+        flex-direction: row;
+        align-items: center;
+        color: $other-dark;
+    }
+    
 }
 
-.input-box, .filled {
+.container-checkbox-component, .filled {
```
Comment (EvertonDuarteGuimaraes) [inline src/assets/scss/checkbox.scss]: It's some reset default styles, why don't put it in early lines? At this point you already have establish some styles to these same classes.

**Review by EvertonDuarteGuimaraes — CHANGES_REQUESTED** [review]
I comment some fixes in the styles class that maybe make it easier to solve the checkbox logic component, i think you should do it.

File: src/assets/scss/checkbox.scss (hunk 2 above)
Comment (RodrigoSoaresNascimento) [reply src/assets/scss/checkbox.scss]: On the site i picked up this checkbox was that way so i left it this way, i don't know what happens if i change.

File: src/assets/scss/checkbox.scss (hunk 1 above)
Comment (RodrigoSoaresNascimento) [reply src/assets/scss/checkbox.scss]: I decrease the name, but the way you want to do I do not understand, so modify to your liking if you want to start from scratch feel free, I understood how fabio did, but as this component is just a just do not know how to apply this.

File: src/components/input-components.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        this._textInput.prop('disabled', true);
+        this.disabled = this._textInput.is(':disabled');
+    }
+
+    setRequired(boolean) {
+        this._textInput.prop('required', boolean);
+        this.required = boolean;
+    }
+
+    isRequired() {
```
Comment (tvps20) [inline src/components/input-components.js]: I will know if it is required by accessing the property and not by the method

File: src/shared/erros/text-input-email.erros.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+export default class InvalidEmail extends Error {
+    constructor(notEmailValid) {
+        super(notEmailValid);
+        this.message = `${notEmailValid} not a valid e-mail address`;
+        this.name = 'EmailInvalid';
+    }
+}
```
Comment (tvps20) [inline src/shared/erros/text-input-email.erros.js:7]: End of file must have a blank line

File: src/pages/components/main.js [hunk 5]
```diff
@@ -1,13 +1,21 @@
...
+LayoutProviderService.inject(page).then(() => {
+    // eslint-disable-next-line no-unused-vars
+    const dropdownGameType = new DropdownComponent('#games');
+    PasswordInputService.initFor('#password');
+    // javascript text-input-box initialization
+    const validateTextInput = new TextInput('.container-text-input-box');
+    validateTextInput.isTextValid();
+    const validateEmailInput = new EmailTextInput('.container-text-input-box');
+    validateEmailInput.isTextValid();
+});
```
Comment (tvps20) [inline src/pages/components/main.js:42]: End of file must have a blank line. Remove lines 18 and 20

File: src/components/text-input.components.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    isTextValid() {
+        if (this._textInput.val() === '') {
+            this._textValid = false;
+            return this._textValid;
+        }
+
+        this._textValid = true;
+        return this._textValid;
+    }
+}
```
Comment (tvps20) [inline src/components/text-input.components.js:27]: End of file must have a blank line

File: src/components/text-input-email.components.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        if (!EmailRegex.test(email)) {
+            this._textValid = false;
+            throw new InvalidEmail(email);
+        } else {
+            this._textInput = this._textInput.val(email);
+            this._textValid = true;
+            return this._textValid;
+        }
+    }
+}
```
Comment (tvps20) [inline src/components/text-input-email.components.js:22]: End of file must have a blank line

File: src/components/input-components.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        if (!this._textInput.prop('required')) {
+            this.required = false;
+            return this.required;
+        }
+
+        this.required = true;
+        return this.required;
+    }
+}
```
Comment (tvps20) [inline src/components/input-components.js:23]: End of file must have a blank line

File: src/components/text-input-email.components.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import TextInput from './text-input.components';
+import InvalidEmail from '../shared/erros/text-input-email.erros';
+
+export default class EmailTextInput extends TextInput {
+    constructor(stringQuery, required = true, _textValid = false, disabled = false) {
+        super(stringQuery, required, _textValid, disabled);
+        this._textInput = $(`${stringQuery} input[type="email"]`);
+    }
+
+    isTextValid(email) {
```
Comment (tvps20) [inline src/components/text-input-email.components.js]: The component already has the value entered in the property, it does not need to pass a value to be validated

**Review by tvps20 — CHANGES_REQUESTED** [review]

File: src/pages/components/main.js (hunk 5 above)
Comment (RodrigoSoaresNascimento) [reply src/pages/components/main.js:42]: Done!

File: src/components/input-components.js (hunk 3 above)
Comment (RodrigoSoaresNascimento) [reply src/components/input-components.js]: fixed and remove this method.

File: src/components/text-input-email.components.js (hunk 9 above)
Comment (RodrigoSoaresNascimento) [reply src/components/text-input-email.components.js]: fixed

Comment (feelsantiago) [inline src/components/text-input-email.components.js]: This can be called just isValid

File: src/components/input-components.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
+// standard class with all states found in the other inputs of this work
+
+export default class Input {
```
Comment (feelsantiago) [inline src/components/input-components.js:3]: Why this input don't have validation?

File: src/components/text-input.components.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+        this._textInput = $(`${stringQuery} input[type="text"]`);
+
+        this._textValid = false;
+    }
+
+    setInput(value) {
+        this._textInput = this._textInput.val(value);
+    }
+
+    isTextValid() {
```
Comment (feelsantiago) [inline src/components/text-input.components.js]: should only be called isValid

**Review by feelsantiago — CHANGES_REQUESTED** [review]

File: src/components/text-input-email.components.js (hunk 9 above)
Comment (RodrigoSoaresNascimento) [reply src/components/text-input-email.components.js]: the more specifies the name of the variable the better for reading, why reduce the name?

File: src/components/text-input.components.js (hunk 11 above)
Comment (RodrigoSoaresNascimento) [reply src/components/text-input.components.js]: Done!

File: src/components/input-components.js (hunk 10 above)
Comment (RodrigoSoaresNascimento) [reply src/components/input-components.js:3]: The validations of this class are in the child classes in both text and email, the idea was to store in them the common properties the validations as required, enable and disable, being necessary to do only the polymorphism in the child classes and adjust the following specifications

File: src/shared/erros/text-input-email.erros.js (hunk 4 above)
Comment (RodrigoSoaresNascimento) [reply src/shared/erros/text-input-email.erros.js:7]: image

File: src/components/text-input.components.js (hunk 6 above)
Comment (RodrigoSoaresNascimento) [reply src/components/text-input.components.js:27]: image

File: src/components/input-components.js (hunk 8 above)
Comment (RodrigoSoaresNascimento) [reply src/components/input-components.js:23]: image

File: src/components/text-input-email.components.js (hunk 7 above)
Comment (RodrigoSoaresNascimento) [reply src/components/text-input-email.components.js:22]: image

**Review by feelsantiago — DISMISSED** [review]

File: src/components/text-input.components.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
+import $ from 'jquery';
+import Input from './input-components';
```
Comment (tvSantiago) [inline src/components/text-input.components.js:2]: Here's a lint error
image

File: src/components/text-input-email.components.js [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+export default class EmailTextInput extends TextInput {
+    constructor(stringQuery, required = true, _textValid = false, disabled = false) {
+        super(stringQuery, required, _textValid, disabled);
+        this._textInput = $(`${stringQuery} input[type="email"]`);
+    }
+
+    isValid() {
+        // Regex with a email form
+        const EmailRegex = /^([a-zA-Z0-9_.+-])+\@(([a-zA-Z0-9-])+\.)+([a-zA-Z0-9]{2,4})+$/;
```
Comment (tvSantiago) [inline src/components/text-input-email.components.js]: Here's a lint error
image

**Review by tvSantiago — CHANGES_REQUESTED** [review]

File: src/components/text-input.components.js (hunk 12 above)
Comment (RodrigoSoaresNascimento) [reply src/components/text-input.components.js:2]: image

Comment (RodrigoSoaresNascimento) [reply src/components/text-input.components.js:2]: The eslint's not picking up here.

File: src/components/text-input-email.components.js (hunk 13 above)
Comment (RodrigoSoaresNascimento) [reply src/components/text-input-email.components.js]: I changed the regex

**Review by tvps20 — DISMISSED** [review]

**Review by tvSantiago — APPROVED** [review]

**Review by tvps20 — APPROVED** [review]

**Review by feelsantiago — APPROVED** [review]

File: src/components/input-components.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
```
Comment (EvertonDuarteGuimaraes) [inline src/components/input-components.js:3]: We already have a class with only enable end disable methods, it don't fit to your class? Can i give you some tips to you fix it?

**Review by EvertonDuarteGuimaraes — CHANGES_REQUESTED** [review]

**Review by EvertonDuarteGuimaraes — APPROVED** [review]

File: src/components/input-components.js (hunk 14 above)
Comment (RodrigoSoaresNascimento) [reply src/components/input-components.js:3]: fixing in the other pr

**Pull request merged**